Enhancement to the virtio queue

jiangliu opened the pull request:
This patch set includes:
1) some minor bugfixes
2) port EVENT_IDX from cloud hypervisor
3) port readable/writable iterator from cloud hypervisor
4) port indirect descriptoer from cloud hypervisor
5) reduce unused dependency on epoll

This is a patch set split out from #13 , and #13 will only includes GuestMemory related changes.

bonzini commented on src/queue.rs:
```diff
@@ -320,6 +320,12 @@ impl<M: GuestAddressSpace> Queue<M> {
         min(self.size, self.max_size)
     }
 
+    /// Reset the queue to a state that is acceptable for a device reset
```
Wrong commit message?

jiangliu replied:
Sorry, fixed:)

bonzini commented on src/queue.rs:
```diff
@@ -142,7 +174,8 @@ impl<M: GuestAddressSpace> DescriptorChain<M> {
 
-    fn checked_new(
+    /// Create a new DescriptorChain instance.
+    pub fn checked_new(
```
Why does this need to be `pub`? Can it be `pub(crate)` instead?

jiangliu replied:
fixed

bonzini commented on src/queue.rs:
```diff
@@ -151,6 +184,52 @@ impl<M: GuestAddressSpace> DescriptorChain<M> {
     }
 
+    /// Create a `DescriptorChain` from the indirect target descriptor table.
+    pub fn new_from_indirect(&self) -> Result<DescriptorChain<M>, Error> {
```
The commit message says `Queue::new_from_indirect`, so that should be fixed.

Would it be possible to add an `Option<DescriptorChain<M>>` field for the indirect descriptor, and add something like

```rust
    if let Some(ref mut indirect) = self.curr_indirect {
        return Some(indirect.next())
    } 
    if self.desc.is_indirect() {
        let indirect_chain = Self::new_from_indirect().ok()?;
        let first = indirect_chain.next();
        self.curr_indirect = Some(indirect_chain);
        return Some(first)
    }

    let curr = self.desc;
    // current code
```

? It can be done in a separate pull request too.

bonzini commented on src/queue.rs:
```diff
@@ -512,14 +512,63 @@ impl<M: GuestAddressSpace> Queue<M> {
+            }
+        };
+        // Note that last_index has no invalid values
+        let last_index: u16 = match mem.read_obj::<u16>(index_addr) {
```
Please use the `integer-atomics` feature of vm-memory instead, and `get_atomic_ref::<u16>`. You can then use normal Rust atomic loads and stores.

jiangliu replied:
Done, please help to review the memory ordering:)

bonzini commented on src/queue.rs:
```diff
@@ -512,14 +512,63 @@ impl<M: GuestAddressSpace> Queue<M> {
+
+        match mem.checked_offset(self.used_ring, (4 + self.actual_size() * 8) as usize) {
+            Some(a) => {
+                mem.write_obj(last_index, a).unwrap();
```
Same here. Note that `get_atomic_ref` takes care of range checking, so you can just use it instead of `checked_offset`.

bonzini commented on src/queue.rs:
```diff
@@ -512,14 +512,63 @@ impl<M: GuestAddressSpace> Queue<M> {
+        }
+
+        // This fence ensures the guest sees the value we've just written.
+        fence(Ordering::Release);
```
This fence is not needed. However, you need a SeqCst fence in the consumer, before it checks the `avail_idx`.

The way QEMU does it is to have a different API, `enable_notification(bool)`. The API does the same thing as your `update_avail_event` if `VIRTIO_RING_F_EVENT_IDX` is not available, but it could also takes care of setting or clearing `VRING_USED_F_NO_NOTIFY` (depending on the bool argument) if event indices are not in use.

With this API, the SeqCst fence is only needed if the bool argument is true.

bonzini commented on src/queue.rs:
```diff
@@ -145,18 +145,13 @@ impl<M: GuestAddressSpace> DescriptorChain<M> {
 
         let desc_size = size_of::<Descriptor>();
-        let desc_head = match mem.checked_offset(desc_table, (index as usize) * desc_size) {
+        let desc_head = match desc_table.checked_add((index as u64) * (desc_size as u64)) {
```
Whenever the code is introduced earlier in the series, can you squash it there?

jiangliu replied:
fixed

bonzini commented on src/queue.rs:
```diff
@@ -112,12 +112,21 @@ impl<M: GuestAddressSpace> DescriptorChain<M> {
         }
 
-        let desc_table_size = size_of::<Descriptor>() * queue_size as usize;
-        let slice = mem.get_slice(desc_table, desc_table_size).ok()?;
```
I agree that we should have a testcase if there is a bug here.

jiangliu replied:
The issue is that, when the whole descriptor table crosses region boundary, it's  legal and should work. With current implementation it fails because the get_slice() works for a single region.

bonzini replied:
@sameo You're confusing `GuestRegion::checked_offset` with `GuestMemory::checked_offset`. But we need a testcase.

jiangliu replied:
Add one test case to cover this case:)

bonzini commented on src/queue.rs:
```diff
@@ -130,6 +130,7 @@ pub struct DescriptorChain<M: GuestAddressSpace> {
 
     /// The current descriptor
     desc: Descriptor,
+    curr_indirect: Option<Box<DescriptorChain<M>>>,
```
Why does this need `Box`?

jiangliu replied:
Otherwise it will fail due to recursive type definition.
```
DescriptorChain {
    curr_indirect: Option<Box<DescriptorChain<M>>>,
}
```

bonzini replied:
Doh, of course.

bonzini commented on src/queue.rs:
```diff
@@ -548,6 +563,74 @@ impl<M: GuestAddressSpace> Queue<M> {
+
+    /// Return the value present in the used_event field of the avail ring.
+    fn get_used_event(&self, mem: &M::M) -> Option<Wrapping<u16>> {
+        // Following code is safe because we have validated the queue. And the `used_index` is a
```
The code is safe not because the queue was validated, but because all accesses are wrapped with `GuestMemory`.

jiangliu replied:
We have used several unchecked_add() below, which depends on queue validation:)

bonzini commented on src/queue.rs:
```diff
@@ -548,6 +563,74 @@ impl<M: GuestAddressSpace> Queue<M> {
+                Wrapping(
+                    s.get_atomic_ref::<AtomicU16>(0)
+                        .unwrap()
+                        .load(Ordering::SeqCst),
```
Never use `SeqCst`, it's really hard to use and can lead to extremely complicated bugs. See [here](https://patchew.org/QEMU/20200407140746.8041-1-pbonzini@redhat.com/20200407140746.8041-6-pbonzini@redhat.com/) for an example.

`SeqCst` fences are okay though.

jiangliu replied:
Thanks for the reference, it's very helpful. Seems that `Ordering::Relaxed` is safe for the EVENT_IDX related operations. 
Yeap, the `SeqCst` is really tricky, a long description in C11 stand:(

bonzini replied:
I suggest you use an `enable_notification` method similar to QEMU's.

bonzini commented on src/queue.rs:
```diff
@@ -548,6 +563,74 @@ impl<M: GuestAddressSpace> Queue<M> {
+    }
+
+    /// Return the value present in the used_event field of the avail ring.
+    fn get_used_event(&self, mem: &M::M) -> Option<Wrapping<u16>> {
```
The `mem` argument should not be necessary in rust-vmm/vm-virtio.

bonzini commented on src/queue.rs:
```diff
@@ -548,6 +563,74 @@ impl<M: GuestAddressSpace> Queue<M> {
+    }
+
+    /// Check whether a notification to the guest is needed.
+    pub fn needs_notification(&mut self, mem: &M::M, used_idx: Wrapping<u16>) -> bool {
```
The `mem` argument should not be necessary here, either.

slp commented:
What's the status of this PR? I see @jiangliu addressed the comments from @bonzini, but looks like the commits haven't been updated. Are we missing a force push here?

alexandruag commented:
By no means should this delay/block anything, but just wanted to say I've started having a better look at the PR as well and I think it we be really nice if we can wrap up some of the ongoing discussions (I'm including other PRs/issues here also), and head towards something like a first version of the queue interfaces/implementations.

slp commented:
> By no means should this delay/block anything, but just wanted to say I've started having a better look at the PR as well and I think it we be really nice if we can wrap up some of the ongoing discussions (I'm including other PRs/issues here also), and head towards something like a first version of the queue interfaces/implementations.

That'd be great. I'm actually trying to put together a common framework for writing vhost-user devices so we can start packaging some implementations in a clean way to ease their way into some distributions. I think we're quite near there, but `vm-virtio::Queue` is missing some functionality that's either critical or very useful for some devices. Indirect descriptors and EVENT_IDX are probably the most important ones (and are covered by this PR). Support for `packed` queues would also be a nice thing to have.

jiangliu commented:
> What's the status of this PR? I see @jiangliu addressed the comments from @bonzini, but looks like the commits haven't been updated. Are we missing a force push here?

I have post the latest code to https://github.com/cloud-hypervisor/vm-virtio/tree/dragonball 
and will repost the latest version.
One ongoing discussion is how to deal with the GuestMemory object.

alexandruag commented on src/queue.rs:
```diff
@@ -517,13 +532,13 @@ impl<M: GuestAddressSpace> Queue<M> {
                 desc_index
             );
-            return;
+            return None;
```
Returning an `Option` from this method doesn't feel quite right, since if we're ever in a position to return `None`, then we've encountered a serious error. I'm not sure what's the best way to handle various `Queue` error conditions (this look like another discussion to be had post this PR), but for now I think this method should either return a `Result<u16>`, or directly  an `u16` that's equal to `self.next_used.0 - 1` (the standard says that, with `VIRTIO_F_EVENT_IDX` negotiated, a notification must be sent when the `used_idx` we've just written to was equal to `used_event`).

alexandruag commented on src/queue.rs:
```diff
@@ -548,6 +563,78 @@ impl<M: GuestAddressSpace> Queue<M> {
+    }
+
+    /// Update avail_event on the used ring with the last index in the avail ring.
+    pub fn update_avail_event(&mut self) {
```
I think this method should be split into an `avail_index()` that returns the available ring index, and then a `set_avail_event()` that writes a specified value to the `avail_event` location in the used ring.

alexandruag commented on src/queue.rs:
```diff
@@ -548,6 +563,78 @@ impl<M: GuestAddressSpace> Queue<M> {
+                    avail_event_slice
+                        .get_atomic_ref::<AtomicU16>(0)
+                        .unwrap()
+                        .store(index, Ordering::Relaxed);
```
AFAIK, `Relaxed` does not generate an appropriate memory fence here. I think we should use `Ordering::Release`.

bonzini replied:
It's okay, the fence (which must be SeqCst) is in the caller.

alexandruag commented on src/queue.rs:
```diff
@@ -548,6 +563,78 @@ impl<M: GuestAddressSpace> Queue<M> {
+    }
+
+    /// Return the value present in the used_event field of the avail ring.
+    fn get_used_event(&self) -> Option<Wrapping<u16>> {
```
The usual naming convention is to skip the `get_` prefix for accessors (`set_` is still used tho for methods that write a value). Also, we should return a `Result` instead of an `Option`, because `None` is not a valid outcome during normal operation.

alexandruag commented on src/queue.rs:
```diff
@@ -548,6 +563,78 @@ impl<M: GuestAddressSpace> Queue<M> {
+                Wrapping(
+                    s.get_atomic_ref::<AtomicU16>(0)
+                        .unwrap()
+                        .load(Ordering::Relaxed),
```
I think we should use `Ordering::Acquire` here.

bonzini replied:
In fact it's not even enough: this function or its caller must have a SeqCst fence at the beginning so that the used ring is filled in before the used event is written.

Please add the fence together with a comment like

    // Complete all the writes in add_used() before reading the event.

alexandruag commented on src/queue.rs:
```diff
@@ -548,6 +563,78 @@ impl<M: GuestAddressSpace> Queue<M> {
+
+        // The VRING_AVAIL_F_NO_INTERRUPT flag isn't supported yet.
+        if self.event_idx {
+            if let Some(old_idx) = self.signalled_used.replace(used_idx) {
```
Hmm, why do we need to keep `self.signalled_used` around? From what I understand, whether we should send a notification or not depends entirely on the values of `used_idx` and `used_event` (the standard specifies a notification should be sent when essentially `used_idx == used_event + 1`, although some device implementations appear to send the notification when the more relaxed `used_idx > used_event` is true).

bonzini replied:
The reason for the more complicated expression is that it allows putting more than one descriptor in the ring and only do a final check for whether to notify. As far as I understand, `signalled_used` is needed to properly handle wrapping arithmetic.

alexandruag commented:
Hi everyone! While there are still some comments to address and alternatives to discuss, it seems best to move forward with everything that's still pending on parallel threads. What do folks think about merging this PR as is, and then adding fixes and improvement proposals on top?

sameo commented:
> Hi everyone! While there are still some comments to address and alternatives to discuss, it seems best to move forward with everything that's still pending on parallel threads. What do folks think about merging this PR as is, and then adding fixes and improvement proposals on top?

Definitely agree, thanks @alexandruag

alexandruag commented:
@jiangliu @bonzini, does this sound good to you?

jiangliu commented:
> @jiangliu @bonzini, does this sound good to you?

There are some fixes for bugs found when integrating the vm-virtio with viritofs device driver.
So updated to the latest version from https://github.com/cloud-hypervisor/vm-virtio/tree/dragonball

bonzini commented on src/queue.rs:
```diff
@@ -614,6 +635,138 @@ impl<M: GuestAddressSpace> Queue<M> {
+
+            // This fence ensures that we observe the latest of virtq_avail once we publish
+            // virtq_used.avail_event/virtq_used.flags.
+            fence(Ordering::AcqRel);
```
This needs to be SeqCst not AcqRel. AcqRel does not order previous stores against future loads.

bonzini reviewed (changes requested):
Still a little work to do on the fences.

bonzini commented:
@jiangliu since you need a few changes on the EVENT_IDX part please also rename `get_used_event` to `used_event` as suggested by Alex.

bonzini commented on src/queue.rs:
```diff
@@ -432,9 +622,149 @@ impl<M: GuestAddressSpace> Queue<M> {
+            // Ensure visibility of virtq_used.idx before sending notification to guest
+            // when the EVENT_IDX feature is disabled, otherwise used_event() will ensure
+            // visibility of virtq_used.idx.
+            fence(Ordering::Release);
```
This is not needed; please remove. Sending a notification (interrupt) is itself a "release" operation and is ordered after previous stores. Just like `used_event()` contains a fence, whatever sends the interrupt (for example a KVM ioctl) will take care of it.

The code is correct without this fence. Note that

```rust
    fence(Ordering::Release);
    fence(Ordering::Acquire);
```

is not the same as `fence(Ordering::SeqCst);`.

jiangliu replied:
Thanks for explanation, removed.

bonzini reviewed (changes requested):
The new fence that was introduced is not necessary.

bonzini commented:
Latest version is okay once CI is green.

alexandruag reviewed (approved):
Thanks everyone! Looks like we can merge this now. I'll write some simple performance tests to help us better understand the impact of various changes, and then create PRs with a couple of proposals. Going forward, I think it's also useful to start an issue to better understand the fence/atomic/sync aspects and decisions we're making as part of this implementation.
